### PR TITLE
Deprecate MorphHelper

### PR DESCRIPTION
**Pull request by daemyung**
These changes bring the loading performance optimization by removing flush and wait.

### CODE REVIEW

**Comment by daemyung** [comment]
This [model](https://github.com/mrdoob/three.js/issues/21636#issuecomment-893520036) and below code are used to measure the performance.
```
auto begin = std::chrono::system_clock::now();
app.resourceLoader->asyncBeginLoad(app.asset);
app.engine->flushAndWait();
auto end = std::chrono::system_clock::now();
std::cout << "compute: " << std::chrono::duration_cast<std::chrono::nanoseconds>(end - begin).count() << std::endl;
```

The loading performance is enhanced around 3.7x(from 531338666 ns to 143684666 ns) on Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz.

File: filament/src/components/RenderableManager.cpp [hunk 1]
```diff
@@ -783,6 +783,12 @@ void RenderableManager::setMorphTargetBufferAt(Instance instance, uint8_t level,
             upcast(morphTargetBuffer), 0, count);
 }
 
+void RenderableManager::setMorphTargetBufferAt(Instance instance, uint8_t level, size_t primitiveIndex,
+        MorphTargetBuffer* morphTargetBuffer) {
+    upcast(this)->setMorphTargetBufferAt(instance, level, primitiveIndex,
+            upcast(morphTargetBuffer), 0, morphTargetBuffer->getVertexCount());
+}
+
```
Comment (pixelflinger) [inline filament/src/components/RenderableManager.cpp:791]: We have a rule that those "trampolines" call the real implementation with the same exact parameters. So you need to create a version without 'count' in the `FRenderableManager`, and have that one call the version with the count.

File: filament/include/filament/RenderableManager.h [hunk 2]
```diff
@@ -488,6 +488,9 @@ class UTILS_PUBLIC RenderableManager : public FilamentAPI {
     void setMorphTargetBufferAt(Instance instance, uint8_t level, size_t primitiveIndex,
             MorphTargetBuffer* morphTargetBuffer, size_t count); //!< \overload
 
+    void setMorphTargetBufferAt(Instance instance, uint8_t level, size_t primitiveIndex,
+            MorphTargetBuffer* morphTargetBuffer); //!< \overload
+
```
Comment (pixelflinger) [inline filament/include/filament/RenderableManager.h:493]: This needs a java version.
Maybe it would make more sense to remove the overload with only a count. Also I think this API change should be in a separate PR.

**Review by romainguy — COMMENTED** [review]
This change will break the Android builds. You need to update the CMake files in android/gltfio-android I believe (and maybe the JNI?)

File: filament/include/filament/RenderableManager.h [hunk 3]
```diff
@@ -488,6 +488,9 @@ class UTILS_PUBLIC RenderableManager : public FilamentAPI {
     void setMorphTargetBufferAt(Instance instance, uint8_t level, size_t primitiveIndex,
             MorphTargetBuffer* morphTargetBuffer, size_t count); //!< \overload
 
+    void setMorphTargetBufferAt(Instance instance, uint8_t level, size_t primitiveIndex,
```
Comment (romainguy) [inline filament/include/filament/RenderableManager.h:491]: This needs to be added to the Java APIs as well.

**Review by pixelflinger — CHANGES_REQUESTED** [review]

File: libs/gltfio/src/ResourceLoader.cpp [hunk 4]
```diff
@@ -503,17 +503,28 @@ bool ResourceLoader::loadResources(FFilamentAsset* asset, bool async) {
             slot.vertexBuffer->setBufferObjectAt(engine, slot.bufferIndex, bo);
             continue;
         }
-        assert(slot.indexBuffer);
-        if (accessor->component_type == cgltf_component_type_r_8u) {
-            const size_t size16 = size * 2;
-            uint16_t* data16 = (uint16_t*) malloc(size16);
-            convertBytesToShorts(data16, data, size);
-            IndexBuffer::BufferDescriptor bd(data16, size16, FREE_CALLBACK);
+        else if (slot.indexBuffer) {
```
Comment (bejado) [inline libs/gltfio/src/ResourceLoader.cpp:506]: nit: move the `else if` so it's on the same line as the `}` (or just delete the `else`)

File: filament/include/filament/RenderableManager.h (hunk 2 above)
Comment (daemyung) [reply filament/include/filament/RenderableManager.h:493]: Ok I see. I will revert this change now and will prepare a new PR for this.

File: filament/include/filament/RenderableManager.h (hunk 3 above)
Comment (daemyung) [reply filament/include/filament/RenderableManager.h:491]: Yes I will create a new PR for these changes.

File: libs/gltfio/src/ResourceLoader.cpp (hunk 4 above)
Comment (daemyung) [reply libs/gltfio/src/ResourceLoader.cpp:506]: Thanks it's fixed. BTW what is nit?

Comment (romainguy) [reply libs/gltfio/src/ResourceLoader.cpp]: nit is short for [nitpick](https://www.google.com/search?q=nitpick&oq=nitpick&aqs=chrome..69i57j69i59l3j69i60l4.742j0j7&sourceid=chrome&ie=UTF-8)

**Comment by daemyung** [comment]
> This change will break the Android builds. You need to update the CMake files in android/gltfio-android I believe (and maybe the JNI?)

@romainguy Yes it's updated.

**Comment by daemyung** [comment]
The loading performance is enhanced around 2.7X(from 292623567 ns to 107575567 ns) on Samsung Exynos 2100 2.91 GHz. But there is no performance enhancement on iOS 7 Plus. I guess iOS 7 Plus has few CPU cores.

**Review by pixelflinger — APPROVED** [review]

**Review by pixelflinger — APPROVED** [review]

**Review by prideout — APPROVED** [review]
